### PR TITLE
[Syntax] [NFC] Remove subclasses of SyntaxData and move validation logic into Syntax

### PR DESCRIPTION
**Pull request by harlanhaskins**
This flattens SyntaxData into a single final class and makes assertions/structural validation an explicit part of Syntax.

It also generalizes the behavior of lazily instantiating children of SyntaxData nodes and simplifies node creation in general.

### CODE REVIEW

**Review by bitjammer — APPROVED** [review]
Code looks good. Very nice remembering to keep the README fresh.

File: include/swift/Syntax/AtomicCache.h [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+public:
+  /// The empty constructor initializes the storage to nullptr.
+  AtomicCache() {}
+
+  /// Gets the value inside the cache, or creates it atomically using the
+  /// provided lambda if it doesn't already exist.
+  RC<T> getOrCreate(std::function<RC<T> ()> Create) const {
+    auto &Ptr = *reinterpret_cast<std::atomic<uintptr_t> *>(&Storage);
+
+    if (Ptr == 0) {
```
Comment (nkcsgexi) [inline include/swift/Syntax/AtomicCache.h]: can we have an early exit here for `if (Ptr`)?

File: include/swift/Syntax/AtomicCache.h [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+      // beats us to caching the child, it'll be non-null, so we would
+      // leave it alone.
+      uintptr_t Expected = 0;
+
+      // Make a RC<T> at RefCount == 1, which we'll try to
+      // atomically swap in.
+      auto Data = Create();
+
+      // Try to swap in raw pointer value.
+      auto SuccessfullySwapped =
```
Comment (nkcsgexi) [inline include/swift/Syntax/AtomicCache.h]: Can we remove `SuccessfullySwapped`? we use it only once.

**Review by nkcsgexi — APPROVED** [review]
Thank you for simplifying the red tree hierarchy! just some minor nitpicks in line.

**Comment by harlanhaskins** [comment]
@swift-ci please smoke test

File: include/swift/Syntax/AtomicCache.h (hunk 2 above)
Comment (bitjammer) [inline include/swift/Syntax/AtomicCache.h]: This was for documentation purposes.

**Comment by harlanhaskins** [comment]
@swift-ci please smoke test

**Comment by harlanhaskins** [comment]
@swift-ci please smoke test

**Comment by harlanhaskins** [comment]
@swift-ci please smoke test and merge

**Comment by harlanhaskins** [comment]
@swift-ci please smoke test and merge

**Pull request merged**